scripts/image-output-extractor/qiskit_docs_notebook_image_extractor/__init__.py

```diff
@@ -0,0 +1,135 @@
+        images_folder = prepare_image_folder(nb_path)
+        nb = nbformat.read(nb_path, 4)
+        print(f"{nb_path}")
+        new_nb, images = extract_images(nb, images_folder)
+        if len(images) == 0:
+            continue
+        for image in images:
+            image.write()
+        nbformat.write(new_nb, nb_path)
+        print(f"  Written notebook and {len(images)} image(s)")
+
+
+def extract_images(
+    nb: nbformat.NotebookNode, image_folder: Path
+) -> tuple[nbformat.NotebookNode, list[Image]]:
+    """
+    Extracts images (converting if necessary) and returns an updated notebook.
+    """
+    images = []
+    for cell_index, cell in enumerate(nb.cells):
+        if cell.cell_type != "code":
+            continue
+        if "outputs" not in cell:
+            continue
+        for index, output in enumerate(cell["outputs"]):
+            if "data" not in output:
+                continue
+            data = output["data"]
+            filestem = Path(image_folder, f"{cell.id}-{index}")
+
+            if svg_data := data.get("image/svg+xml", None):
+                image = SvgImage(filepath=filestem.with_suffix(".svg"), data=svg_data)
+            elif png_data := data.get("image/png", None):
+                png_image = RasterImage(
+                    filepath=filestem.with_suffix(".png"),
+                    data=base64.b64decode(png_data),
+                )
+                print(f"  - Converting PNG output for cell {cell_index}")
+                image = convert_to_avif(png_image)
+            else:
+                continue
+
+            data["text/plain"] = mdx_component(image.filepath)
+            # Delete all image outputs now we've converted one.
+            # An output can have many different representations (e.g. text,
+            # html, image), including many image representations in different
+            # formats. We only want to keep one image representation, so we ignore
+            # the rest.
+            for datatype in ["png", "jpeg", "svg+xml"]:
+                data.pop(f"image/{datatype}", None)
+            images.append(image)
+    return nb, images
+
+
+def mdx_component(image_path: Path) -> str:
+    return f'<Image src="/{image_path.relative_to("public")}" alt="Output of the previous code cell" />'
+
+
+def convert_to_avif(image: RasterImage) -> RasterImage:
+    """
+    Pipe image through ImageMagick subprocess to convert to AVIF.
+    """
+    new_path = image.filepath.with_suffix(".avif")
+    imagemagick = Popen(["magick", "-", "avif:-"], stdout=PIPE, stderr=PIPE, stdin=PIPE)
+    (new_data, _stderr) = imagemagick.communicate(input=image.data)
+    return RasterImage(filepath=new_path, data=new_data)
+
+
+def prepare_image_folder(nb_path: Path) -> Path:
+    """
+    Determine the appropriate output folder for the extracted images, and ensure it exists and is empty.
+    """
+    images_folder = Path(
+        "public",
+        nb_path.parts[0],
```

Comment:
It'd be helpful to have a comment explaining what the parts correspond to, e.g. `/my_path/foo.ipynb` -> `my_path`